tycho-indexer/src/extractor/dynamic_contract_indexer/rpc_metadata_provider.rs

```diff
@@ -0,0 +1,664 @@
+#![allow(unused_variables)] // TODO: Remove this once the provider is implemented
+#![allow(dead_code)] // TODO: Remove this once the provider is implemented
+
+use std::{collections::HashMap, sync::Arc};
+
+use reqwest::Client;
+use serde_json::{json, Value};
+use tonic::async_trait;
+use tracing::warn;
+
+use crate::extractor::dynamic_contract_indexer::component_metadata::{
+    DeduplicationId, MetadataError, RequestProvider, RequestTransport, RpcTransport,
+};
+
+pub struct RPCMetadataProvider {
+    client: Arc<Client>,
+    batch_size_limit: usize,
+}
+
+impl RPCMetadataProvider {
+    pub fn new(batch_size_limit: usize) -> Self {
+        Self { client: Arc::new(Client::new()), batch_size_limit }
+    }
+}
+
+#[async_trait]
+impl RequestProvider for RPCMetadataProvider {
+    async fn execute_batch(
+        &self,
+        requests: &[Box<dyn RequestTransport>],
+    ) -> Vec<(DeduplicationId, Result<Value, MetadataError>)> {
+        let mut results: HashMap<DeduplicationId, Result<Value, MetadataError>> =
+            HashMap::with_capacity(requests.len());
+
+        let batches = self.group_requests(requests, self.batch_size_limit);
+
+        for batch in batches {
+            let mut rpc_requests = Vec::new();
+
+            for request in batch {
+                match request
+                    .as_any()
+                    .downcast_ref::<RpcTransport>()
+                {
+                    Some(rpc_transport) => rpc_requests.push(rpc_transport.clone()),
+                    None => {
+                        results.insert(
+                            request.deduplication_id(),
+                            Err(MetadataError::ProviderFailed(
+                                "Invalid rpc request transport".into(),
+                            )),
+                        );
+                    }
+                }
+            }
+
+            if rpc_requests.is_empty() {
+                continue;
+            }
+
+            let endpoint = rpc_requests[0].endpoint.clone();
+            let all_same_endpoint = rpc_requests
+                .iter()
+                .all(|rpc| rpc.endpoint == endpoint);
+
+            // If the requests are not all to the same endpoint, we can't batch them
+            // and we need to insert errors for each request
+            // TODO: if the final design still include endpoint inside RpcTransport, we need to
```

Comment:
Yeah I want to keep the possibility of having different RPC entrypoints but I struggle to see an use case rn. So, for now, I'm fine with raising this error, makes it easier to spot that we have a configuration mistake